src/libraries/System.IO.Compression.Zstandard/src/System/IO/Compression/ZstandardStream.Compress.cs

```diff
@@ -0,0 +1,322 @@
+                        lastResult = _encoder.Compress(buffer, output, out bytesConsumed, out bytesWritten, isFinalBlock);
+                    }
+
+                    if (lastResult == OperationStatus.InvalidData)
+                    {
+                        throw new InvalidDataException(SR.ZstandardStream_Compress_InvalidData);
+                    }
+                    if (bytesWritten > 0)
+                    {
+                        _stream.Write(output.Slice(0, bytesWritten));
+                    }
+                    if (bytesConsumed > 0)
+                    {
+                        buffer = buffer.Slice(bytesConsumed);
+                    }
+                }
+            }
+            finally
+            {
+                if (checkActiveRWOps)
+                {
+                    EndRWOperation();
+                }
+            }
+        }
+
+        private async ValueTask WriteCoreAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken, bool isFinalBlock = false, bool flush = false, bool checkActiveRWOps = true)
+        {
+            if (_mode != CompressionMode.Compress)
+            {
+                throw new InvalidOperationException(SR.CannotWriteToDecompressionStream);
+            }
+            EnsureNotDisposed();
+            Debug.Assert(_encoder != null);
+
+            try
+            {
+                if (checkActiveRWOps)
+                {
+                    BeginRWOperation();
+                }
+
+                OperationStatus lastResult = OperationStatus.DestinationTooSmall;
+
+                // we don't need extra tracking of ArrayBuffer for write operations as
+                // we propagate the entire result downstream, so just grab the memory
+                Memory<byte> output = _buffer.AvailableMemory;
+                Debug.Assert(!output.IsEmpty, "Internal buffer should be initialized to non-zero size");
+
+                while (lastResult == OperationStatus.DestinationTooSmall)
+                {
+                    int bytesWritten;
+                    int bytesConsumed;
+
+                    if (flush)
+                    {
+                        Debug.Assert(buffer.Length == 0);
+                        bytesConsumed = 0;
+                        lastResult = _encoder.Flush(output.Span, out bytesWritten);
+                    }
+                    else
+                    {
+                        lastResult = _encoder.Compress(buffer.Span, output.Span, out bytesConsumed, out bytesWritten, isFinalBlock);
+                    }
+
+                    if (lastResult == OperationStatus.InvalidData)
+                    {
+                        throw new InvalidOperationException(SR.ZstandardStream_Compress_InvalidData);
+                    }
+                    if (bytesWritten > 0)
+                    {
+                        await _stream.WriteAsync(output.Slice(0, bytesWritten), cancellationToken).ConfigureAwait(false);
```

Comment:
This is the first time the WriteAsync call will go async, which means the first compression or flush will happen in the sync portion of the method - is that what we want?

Reply:
I don't see us yielding to threadpool before de/compression in compression stream implementations.

The buffer is small (8kB) so there should not be a long delay. I would prefer not yielding unless we have reports that it affects users in a negative way, and if we decide to make a change, we should do it for all compression streams at once.

Reply:
let's continue the discussion in the similar thread in Decompress.